### DISCUSSION TITLE
Example service installation

### DISCUSSION THREAD

**Discussion by MarkStega** [wixtoolset/issues#7326] [Questions] [answered]
BACKGROUND:
I successfully migrated my legacy app (Net Framework 4.8) installer from Wix3 to Wix4. The application is composed of a number of Windows Services.

I now am trying to create an installer for my 'modern' app (.Net 7.0) which is also a series of Windows Services or Linux applications. The Linux side is handled with the creation of docker images. I want to use Wix4 for the Windows services.

ISSUE:
I basically took my .wxs source from the legacy app and am trying to use the same structure for the .Net services.

I am receiving two errors (repeated twice) when I build the installer.
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	WIX0204	ICE30: The target file 'epomgjtf.dll|CommunicationService.dll' is installed in '[ProgramFiles64Folder]\enfrobmm\ybaf43nb\' by two different components on an SFN system: 'ComponentIdCS' and 'cmpXj_C7WEfEvNi45vkrPKWqiUyRzc'. This breaks component reference counting.	WiX	C:\Solutions\OHI\Optimiser\WiX\obj\x64\SERVICE\_CommunicationService_dir.wxs	28	

and

Severity	Code	Description	Project	File	Line	Suppression State
Error	WIX0204	ICE30: The target file 'epomgjtf.dll|CommunicationService.dll' is installed in '[ProgramFiles64Folder]\enfrobmm\ybaf43nb\' by two different components on an SFN system: 'ComponentIdCS' and 'cmpXj_C7WEfEvNi45vkrPKWqiUyRzc'. This breaks component reference counting.	WiX	C:\Solutions\OHI\Optimiser\WiX\Product.wxs	80	
```
Is there an example wxs source that I can use as a model for my wxs?

**Comment by BMurri** [comment, marked as answer]
What is the dfiference between the two components? Does one install the service and the other doesn't? You need to either remove one of the components or remove the file from one of them since any given file should only ever be in just one component.

**Reply by MarkStega** [reply]
Here is what I have in the wixproj:
```
  <ItemGroup>
    <ProjectReference Include="..\CommunicationService\CommunicationService.csproj">
      <Name>CommunicationService</Name>
      <Project>{1bc2158b-eda0-429c-91ac-5459de7b895a}</Project>
      <Private>True</Private>
      <DoNotHarvest>True</DoNotHarvest>
      <RefProjectOutputGroups>Binaries;Content;Satellites</RefProjectOutputGroups>
      <RefTargetDir>INSTALLFOLDER</RefTargetDir>
    </ProjectReference>
</ItemGroup>
```
and here is the relevant wxs content:
```
<?define ServiceNameCS = "Optimi$er.CommunicationService" ?>
<?define ServiceDescriptorCS = "Send and receives email, fax, and text messages" ?>
<?define ServiceTargetFileNameCS = $(var.CommunicationService.TargetFileName)?>
<?define ServiceQualifiedTargetFileNameCS = "$(var.CommunicationService.TargetDir)$(var.CommunicationService.TargetFileName)"?>
...
		<DirectoryRef Id="ApplicationRootDirectoryCS">
			<Component Id="ComponentIdCS"
					   Guid="{19CD5363-7065-4CB4-A863-D4CC6B82362B}">
				<File Id="ServiceTargetFileNameCS"
					  Name="$(var.ServiceTargetFileNameCS)"
					  Source="$(var.ServiceQualifiedTargetFileNameCS)" />
				<ServiceInstall Id="ServiceInstallerCS"
								Type="ownProcess"
								Name="$(var.ServiceNameCS)"
								DisplayName="$(var.ServiceNameCS)"
								Description="$(var.ServiceDescriptorCS)"
								Start="demand"
								Account="LocalSystem"
								ErrorControl="normal" />
				<ServiceControl Id="StartControlCS"
								Stop="both"
								Remove="both"
								Name="$(var.ServiceNameCS)"
								Wait="yes" />
			</Component>
		</DirectoryRef>
...
		<Feature Id="ServiceId1" Title="$(var.ServiceNameCS)" Level="1">
			<ComponentRef Id="ComponentIdCS" />
			<ComponentGroupRef Id="CommunicationService" />
		</Feature>
...
		<StandardDirectory Id="ProgramFiles6432Folder">
			<!-- Create a folder inside program files called Optimium Health -->
			<Directory Id="ROOTDIRECTORY" Name="$(var.Manufacturer)">
				<!-- Create a folder inside Optimium Health for each service -->
				<Directory Id="ApplicationRootDirectoryCS" Name="$(var.ServiceNameCS)" />
			</Directory>
		</StandardDirectory>
```

**Reply by robmen** [reply]
It turns out yesterday's [Deployment Dojo episode 25 was about installing services](https://youtube.com/live/vZRZeDOTPZQ). Perhaps that would be helpful for you.

**Reply by MarkStega** [reply]
@robmen 
Thanks for the reference, I had no idea that the deployment dojo series existed; Since I was successfully installing services for many years using WiX3 and my legacy Net 4.8 series of apps it was good to see that the pattern I was using was valid.

@BMurri 
I did find the issue that was giving me the error:

I was using a preprocessor variable definition of
```
<?define ServiceTargetFileNameCS = $(var.CommunicationService.TargetFileName)?>
```
and the problem is that for a .Net application the "$(var.CommunicationService.TargetFileName)" resolves to CommunicationService.dll rather than to CommunicationService.exe; By hard coding the target file name the installer gets created as expected.

So is this an expected new behavior for the TargetFileName and, if so, is there a new variable that does point to the executable?

**Reply by BMurri** [reply]
Because the "new" .NET (built on the former .Net Core) is multi-platform, executables are now built as .dll files, with a platform-specific runner added. Thus, `TargetFileName` points to the built dll. I think I saw an undocumented msbuild item, but it's always the `TargetFileName` with the extension changed to `.exe`.

**Reply by MarkStega** [reply]
@BMurri

Thanks, not what I expected, but an easy 'work around' exists
